site/hooks/useWeb3/index.tsx

```diff
@@ -0,0 +1,140 @@
+import { useEffect, useReducer, useCallback } from "react";
+import WalletConnectProvider from "@walletconnect/web3-provider";
+import { urls } from "@klimadao/lib/constants";
+import { ethers } from "ethers";
+import WalletLink from "walletlink";
+import Web3Modal from "web3modal";
+
+import {
+  Web3ProviderState,
+  Web3Action,
+  web3InitialState,
+  web3Reducer,
+} from "./reducers";
+
+const providerOptions = {
+  walletconnect: {
+    package: WalletConnectProvider,
+    options: {
+      rpc: { 137: urls.polygonMainnetRpc },
+    },
+  },
+  walletlink: {
+    package: WalletLink,
+    options: {
+      appName: "Official KlimaDAO App",
+      rpc: urls.polygonMainnetRpc,
+      chainId: 137,
+      appLogoUrl: null,
+      darkMode: false,
+    },
+  },
+};
+
+let web3Modal: Web3Modal | null;
+if (typeof window !== "undefined") {
+  web3Modal = new Web3Modal({
+    cacheProvider: true,
+    providerOptions,
+  });
+}
+
+export const useWeb3 = () => {
+  const [state, dispatch] = useReducer(web3Reducer, web3InitialState);
+  const { provider, web3Provider, address, network } = state;
+
+  const connect = useCallback(async () => {
+    if (web3Modal) {
+      try {
+        const provider = await web3Modal.connect();
+        const web3Provider = new ethers.providers.Web3Provider(provider);
+        const signer = web3Provider.getSigner();
+        const address = await signer.getAddress();
+        const network = await web3Provider.getNetwork();
+
+        dispatch({
+          type: "SET_WEB3_PROVIDER",
+          provider,
+          web3Provider,
+          address,
+          network,
+        } as Web3Action);
+      } catch (e) {
+        console.log("Connect error:", e);
+      }
+    } else {
+      console.error("No Web3Modal");
+    }
+  }, []);
+
+  const disconnect = useCallback(async () => {
+    if (web3Modal) {
+      web3Modal.clearCachedProvider();
+      if (provider?.disconnect && typeof provider.disconnect === "function") {
+        await provider.disconnect();
```

Comment:
we need to reset the entire app state when they disconnect (remove all balances etc).
so that's why I was calling window.reload on the old one

Reply:
Gotcha! What do you think of defining a new disconnect function in `app` that handles the page reload?